Feat: Support Dapp for Starcoin

wk3368 opened the pull request:
* 6728ff77 fix: format by yarn lint --fix
* 70485ce4 fix: support starmask_getProviderState, gasPrice should not be 0.000000001, add tokensChangedTo in fetchFeeInfo response
* 902bb4f1 feat: decode txParams.data to display dataName and dataParamsStr on UI's confirmTransactionPage
* 14c74184 feat: Support getNextNonce for stc
* ad2eaae0 feat: onekey hardware wallet support personal sign
* 005adff7 fix: support txParams.data
* b758c476 fix: gasPrice should be 1
* 57bf3379 feat: support scriptFunction from dapp
* b73f8002 feat: add wallet_getPermissions
* f87d5f7c feat: add proxyJsonRPCCall for starcoin
* a3e63d31 feat: Add support for methods: stc_requestAccounts  &&  stc_accounts
* 5860f2f3 feat: setup ProviderApiStarcoin

wk3368 committed:
fix: /app-monorepo/node_modules/keytar/build/Release/keytar.node (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')

wk3368 committed:
feat: setup ProviderApiStarcoin

wk3368 committed:
feat: Add support for methods: stc_requestAccounts  &&  stc_accounts

wk3368 committed:
feat: add proxyJsonRPCCall for starcoin

wk3368 committed:
feat: add wallet_getPermissions

wk3368 committed:
feat: support scriptFunction from dapp

wk3368 committed:
fix: delete unused  signTransaction, the one in vaults/keyring/keyringSoftwareBase works in runtime

wk3368 committed:
fix: gasPrice should be 1

wk3368 committed:
fix: revert delete unused signTransaction, it is used by onekey hardware wallet

wk3368 committed:
fix: support txParams.data

wk3368 committed:
feat: onekey hardware wallet support personal sign

wk3368 committed:
feat: Support getNextNonce for stc

wk3368 committed:
feat: decode txParams.data to display dataName and dataParamsStr on UI's confirmTransactionPage

sidmorizon commented on packages/desktop/src-electron/preload.ts:
```diff
@@ -65,10 +65,12 @@ const desktopApi = {
   secureSetItemAsync(key: string, value: string) {
-    return keytar.setPassword('OneKey', key, value);
+    return Promise.resolve();
+    // return keytar.setPassword('OneKey', key, value);
```
线上的 Desktop 包能运行起来吗
https://onekey.so/download

这里的一些临时注释的代码需要还原

wk3368 replied:
嗯。线上的desktop在本地可以跑。

sidmorizon commented on packages/engine/src/vaults/VaultBase.ts:
```diff
@@ -439,4 +441,37 @@ export abstract class VaultBase extends VaultBaseChainOnly {
     return IDecodedTxDirection.OTHER;
   }
+
+  async getNextNonce(
```
主分支上，这个方法有一些修改，需要重新从 evm 里挪过来

sidmorizon commented on packages/engine/src/vaults/impl/evm/Vault.ts:
```diff
@@ -373,13 +373,13 @@ export default class Vault extends VaultBase {
       );
       if (!token) {
-        throw new Error(`Token not found: ${transferInfo.token as string}`);
+        throw new Error(`Token not found: ${ transferInfo.token as string }`);
```
在本地跑一遍 yarn lint --fix 再 push 代码，这样不会导致格式化的不统一

wk3368 committed:
fix: support starmask_getProviderState, gasPrice should not be 0.000000001, add tokensChangedTo in fetchFeeInfo response

wk3368 committed:
fix: format by yarn lint --fix

wk3368 committed:
fix: Merge branch 'onekey' into providerApiStarcoin

# Conflicts:
#	packages/desktop/src-electron/preload.ts
#	packages/engine/src/vaults/impl/evm/Vault.ts
#	packages/engine/src/vaults/impl/stc/Vault.ts
#	packages/kit/src/views/Send/SendConfirmViews/SendConfirmModal.tsx

wk3368 committed:
fix: /app-monorepo/node_modules/keytar/build/Release/keytar.node (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')

wk3368 committed:
fix: add starcoin in isNetworkNotMatched

wk3368 committed:
fix: Move getNextNonce into ValtBase

wk3368 committed:
fix: format by yarn lint --fix

sidmorizon commented on packages/desktop/src-electron/preload.ts:
```diff
@@ -1,6 +1,6 @@
 import { ipcRenderer } from 'electron';
 import logger from 'electron-log';
-import keytar from 'keytar';
+// import keytar from 'keytar';
```
这个文件的改动不能提交上来

wk3368 committed:
fix: reverse changes in preload.ts

sidmorizon commented on packages/engine/src/vaults/impl/stc/Vault.ts:
```diff
@@ -82,29 +100,45 @@ export default class Vault extends VaultBase {
+        },
+      };
+    } else {
+      throw new NotImplemented();
```
这里不要抛异常，其他情况也属于 IDecodedTxActionType.TRANSACTION ，只是 data 为空而已，这样不会因为抛错阻止用户正常交易

sidmorizon commented on packages/engine/src/vaults/impl/stc/Vault.ts:
```diff
@@ -113,7 +147,16 @@ export default class Vault extends VaultBase {
       },
       extraInfo: null,
+      encodedTx,
+      ...(encodedTx.data
```
直接写 payload: {
              data: encodedTx.data,
            }

即可，不需要判断为空

sidmorizon commented on packages/engine/src/vaults/impl/stc/Vault.ts:
```diff
@@ -196,7 +256,7 @@ export default class Vault extends VaultBase {
     );
 
-    return { ...unsignedTx, encodedTx };
+    return unsignedTx as IUnsignedTxPro;
```
这里不能简单的 as，而是通过 { ...unsignedTx, encodedTx } 需要把 encodedTx 传回去的，不然前端会报错

sidmorizon commented on packages/engine/src/vaults/impl/stc/Vault.ts:
```diff
@@ -205,8 +265,7 @@ export default class Vault extends VaultBase {
       ...encodedTx,
-      nonce: 1,
-      gasPrice: '1',
+      gasPrice: '0.000000001',
```
encodedTx 这里的 price 是整数值 1

wk3368 replied:
这个地方必须是0.000000001

否则在dry_run_raw 获取gas的时候，rpc返回会报错

<img width="2190" alt="image" src="https://user-images.githubusercontent.com/171268/180147910-74f56779-200d-4e3a-93e6-771ed0d4b0b7.png">

显示的 Fee 永远是写死的默认值：
0.000127845 TSTC

wk3368 replied:
<img width="2739" alt="image" src="https://user-images.githubusercontent.com/171268/180148448-b8744a7f-dfac-4960-8b09-406877382b38.png">

正常的dry_run_raw的结果，应该是 
Fee
0.000105547 TSTC

wk3368 replied:
你们之前的就是错的，dry_run_raw取预估的gas都是失败的，提交到链上执行是ok的。
只是报错你们没有发现而已。

sidmorizon commented on packages/engine/src/vaults/VaultBase.ts:
```diff
@@ -8,12 +8,20 @@ import {
 import { PartialTokenInfo } from '@onekeyfe/blockchain-libs/dist/types/provider';
```
packages/components/svg/outline/expand.svg
这个文件删除要 reset ，不能提交这个文件的删除

sidmorizon commented on packages/engine/src/vaults/impl/stc/Vault.ts:
```diff
@@ -54,25 +63,34 @@ export default class Vault extends VaultBase {
   };
 
   decodedTxToLegacy(decodedTx: IDecodedTx): Promise<IDecodedTxLegacy> {
```
这个方法可以删除了，后面我来清理下

wk3368 replied:
我现在不能删，删了就报错，等你清理的时候统一删除吧。

sidmorizon commented on packages/kit/src/background/providers/ProviderApiStarcoin.ts:
```diff
@@ -0,0 +1,592 @@
+  @providerApiMethod()
+  stc_coinbase(request: IJsBridgeMessagePayload) {
+    // TODO some different with stc_accounts, check metamask code source
+    return this.stc_accounts(request);
```
这个方法之前 evm 错了，最新代码已经调整，返回的是 string 不是 数组

sidmorizon commented on packages/kit/src/background/providers/ProviderApiStarcoin.ts:
```diff
@@ -0,0 +1,592 @@
+   */
+  @providerApiMethod()
+  stc_signTransaction(req: IJsBridgeMessagePayload, ...params: string[]) {
+    if (params[1].length === 66 || params[1].length === 67) {
```
这里直接     throw web3Errors.rpc.methodNotSupported(); 即可，之前 evm 也没支持，代码是 copy 错误的

wk3368 committed:
fix: some changes according to the  code review comments

wk3368 committed:
fix: upgrade dependencies version

Commit:
Merge branch 'onekey' into providerApiStarcoin
